README.md
# wercker - your new favorite dev tool
[![wercker status](https://app.wercker.com/status/febe6e1691586f99d20eb79c6b706aaa/s/master "wercker status")](https://app.wercker.com/project/bykey/febe6e1691586f99d20eb79c6b706aaa)

This is the project for `wercker`, the command-line tool that powers
all the build and deploy jobs for [wercker.com](http://wercker.com), it
runs on your local machine with the help of Docker.

Wercker is designed to increase developer velocity by enabling users to run
and automate their tests and builds, leveraging Docker containers to
provide development environments for multi-service architectures.

Note: While the Wercker team has extensive experience with open source, this
project has been internal for some time and may have some rough edges. We'll
be actively prettying up the codebase and the project, but we wanted to
release it to you as early as possible.

The master branch may be in a broken or unstable state during development.
It is recommended that you download `wercker` through
[the CLI section](http://wercker.com/cli/) on our website, if you're not
contributing to the code base.

## Building wercker

`wercker` is built using Go version 1.5 or greater. If you don't have it
already, you can get it from the
[official download page](https://golang.org/dl/). Once you go installed, set
up your go environment by
[using this guide](https://golang.org/doc/code.html#Organization)

Next, you'll need `govendor` to install the golang dependencies. You can do
this by running:
```
  go get github.com/kardianos/govendor
```

In Go 1.5 you'll need the vendor experiment enabled, so make sure to export
`GO15VENDOREXPERIMENT=1` in your shell (see [Go 1.5 Vendor Experiment](https://docs.google.com/document/d/1Bz5-UB7g2uPBdOx-rw5t9MxJwkfpx90cqG9AFL0JAYo/edit))

In your git checkout ($GOPATH/src/github.com/wercker/wercker), run:
```
   govendor sync
```

This command should download the appropriate dependencies.

Once all that is setup, you should be able to run `go build` and get a working
executable named `wercker`.

Once you've got a working Docker environment, running
```
  ./wercker build
```

should go through the entire build and testing process.

Note: this is the bare minimum to build and contribute to the code base. If you
do not have a local Docker environment you will not be able to run and test
`wercker` properly. You can follow [this guide](https://docs.docker.com/engine/installation/) to install Docker on your machine.

## Reporting Bugs

If you are experiencing bugs running wercker locally, please create an issue
containing the following:

- Which OS are you using?
- Which Docker environment are you using? (Boot2docker, custom, etc)
- Create a gist containing the following information:
  - The entire log when running wercker with the `--debug` log. (ie. `wercker --debug build`)
  - The wercker.yml file that causes the issues.

Please don't file any issue dealing with the usage of steps or unexpected behavior on hosted wercker.

If you are experiencing issues running builds or deploys on hosted wercker,
please do the following:

Try running the build again to see if the error keeps occurring. If it does, turn
on support for the application, and create an issue with the following
information:

- The application owner and application name.
- The ID of the build or deploy that failed.

## Contact

Join us in our slack room: [![Slack Status](http://werckerpublicslack.herokuapp.com/badge.svg)](http://slack.wercker.com)

## License

`wercker` is under the Apache 2.0 license. See the [LICENSE](LICENSE) file for details.


changelog.md
## unreleased

- Changes for robust error handling and reporting in docker-push (#387)
- Replace docker links with docker network (#382)

## v1.0.1201 (2018-04-16)

- Update azure client to allow docker-push in all regions (#381)

## v1.0.1196(2018-04-11)

- Fixes and additional properties for internal/docker-build step (#372)

## v1.0.1195(2018-04-10)

- Fix for correctly inferring regsitry and repoistory from step inputs (#375) 
- Fix "go build" and "wercker build" on golang 1.10 (#374)

## v1.0.1189(2018-04-04)

- Fix status reporting for docker push (#371)

## v1.0.1183 (2018-03-29)

- New docker-build step and enhanded docker-push step (#362)

## v1.0.1153 (2018-02-27)

- Remove Keen dependencies (#354)

## v1.0.1062 (2017-11-28)

- Default docker hub push to registry V2 (#348)

## v1.0.1049 (2017-11-15)

- Update dependencies, as a result of `Sirupsen/logrus` -> `sirupsen/logrus` (#333)
- Add a Docker subcommand (#335)
- Ensure repository names are always lowercase (#338)
- Support for the new step manifest format (#343)

## v1.0.965 (2017-08-23)

- Change compilation in separate wercker steps (#331)
- Add retry and exponential backoff for fetching step metadata and step tarball
  (#330)
- Add flag to delete Docker image after pushing it to a registry (#327)
- Use wercker registry for wercker-init (#334)

## v1.0.938 (2017-07-27)

- Some nice additions to the way we do the docker push and test (#320)
- Fix env var loading order (#314, #315, #317)
- Fix internal/watchstep (#312)
- Add env option to docker-scratch-push (#295)
- Allow relative paths for file:// targets in dev mode (#296)
- Better control limiting memory on run containers, when using
  services gives the services a 25% of the total memory to split
  amongst themselves, defaults to no limits (#299)
- Automatically detect bash or sh for containers by default,
  defaulting to bash if it is there (#301)
- Fix a small bug when doing local deploys and using a working-dir other
  than .wercker (#301)

## v1.0.758 (2017-01-27)

- Add Azure Registry support (#275)
- Explicitly chmods the basepath / source path to a+rx
- Removes the explicit clear after launching a shell (#257)
- Fix `wercker doc` and update `./Documentation/*` (#260)

## v1.0.643 (2016-10-05)

- Remove google as default container DNS (#245)
- Update to compiling with go 1.7

## v1.0.629 (2016-09-21)

- Add additional output when storing artifacts (#207)
- Fix longer (2+) chains of runs that have source-dir specified (#151)
- Output more descriptive error message when setup environment fails (#230)
- Allow use of an "ignore-file" yaml directive that parse the gitignore syntax
  (#240)

## v1.0.560 (2016-07-14)

- Fix internal/docker-scratch-push for Docker 1.10+

## v1.0.547 (2016-07-01)

- Add checkpointing and base-path (#123)
- Support for registry v2 (#131)
- Mount volumes in the container from different local paths (#134)
- Only push tags that were defined in the wercker.yml (#142)
- wercker is now using govendor (#146)
- Display raw config, before parsing it (#149)
- Allow multiple services with the same images (#159)
- Add exposed-ports (#161)
- Fix run, build and deploy urls (#163)

## 2016.03.11

### Features

- Moves the working path to default to `.wercker` and removes the flags
  for configuring the other paths
- Adds a symlink `.wercker/latest` for referring to your latest build, and
  a `.wercker/latest_deploy` for referring to your latest deploy
- Make the --artifacts work better locally, making your build's artifacts
  easily available under .wercker/latest/output
- Automatically use the contents of `.wercker/latest/output` when running a
  `wercker deploy` without specifying a target
- When running `wercker deploy` if the specified target does not container a
  wercker.yml file, attempt to use the one in the current directory.
- Allow settings multiple tags at a time when doing `internal/docker-push`
- Check for and allow unix:///var/run/docker.sock on non-linux hosts


### Bug Fixes

- Deal with symlinks significantly better
- Respect --docker-local when using `internal/docker-push` (don't push)
- Allow images to be pulled by nested local services (removes
  implicit --docker-local)
- Workaround a docker issue related to not fully consuming the result of a
  CopyFromContainer API call (when we exported a cache that was more than our
  limit of 1GB we'd just drop it, and docker would hang)
- Remove pipeline ID tag set by `internal/docker-push`


## 2016.02.10

### Features

- Allow users to mount local volumes to their wercker build containers, specified by a list of `volumes` underneath box in the wercker.yml file. Must have `--enable-volumes` flag set in order to run.
- Check to see if config from wercker.yml is empty
- Adds changelog

### Bug fixes

- Fixes to the shellstep implementation


cmd/flags.go
//   Copyright © 2016, 2018, Oracle and/or its affiliates.  All rights reserved.
//
//   Licensed under the Apache License, Version 2.0 (the "License");
//   you may not use this file except in compliance with the License.
//   You may obtain a copy of the License at
//
//       http://www.apache.org/licenses/LICENSE-2.0
//
//   Unless required by applicable law or agreed to in writing, software
//   distributed under the License is distributed on an "AS IS" BASIS,
//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//   See the License for the specific language governing permissions and
//   limitations under the License.

package cmd

import (
	"github.com/codegangsta/cli"
	"github.com/wercker/wercker/core"
)

// Flags for setting these options from the CLI
var (
	// These flags tell us where to go for operations
	EndpointFlags = []cli.Flag{
		// deprecated
		cli.StringFlag{Name: "wercker-endpoint", Value: "", Usage: "Deprecated.", Hidden: true},
		cli.StringFlag{Name: "base-url", Value: core.DEFAULT_BASE_URL, Usage: "Base url for the wercker app.", Hidden: true},
		cli.StringFlag{Name: "steps-registry", Value: "https://steps.wercker.com", EnvVar: "STEPS_REGISTRY", Usage: "Endpoint for the steps registry", Hidden: true},
	}

	// These flags let us auth to wercker services
	AuthFlags = []cli.Flag{
		cli.StringFlag{Name: "auth-token", Usage: "Authentication token to use."},
		cli.StringFlag{Name: "auth-token-store", Value: "~/.wercker/token", Usage: "Where to store the token after a login.", Hidden: true},
	}

	DockerFlags = []cli.Flag{
		cli.StringFlag{Name: "docker-host", Value: "", Usage: "Docker api endpoint.", EnvVar: "DOCKER_HOST"},
		cli.StringFlag{Name: "docker-tls-verify", Value: "0", Usage: "Docker api tls verify.", EnvVar: "DOCKER_TLS_VERIFY"},
		cli.StringFlag{Name: "docker-cert-path", Value: "", Usage: "Docker api cert path.", EnvVar: "DOCKER_CERT_PATH"},
		cli.StringSliceFlag{Name: "docker-dns", Value: &cli.StringSlice{}, Usage: "Docker DNS server.", EnvVar: "DOCKER_DNS", Hidden: true},
		cli.BoolFlag{Name: "docker-local", Usage: "Don't interact with remote repositories"},
		cli.StringFlag{Name: "checkpoint", Value: "", Usage: "Skip to the next step after a recent build checkpoint."},
		cli.IntFlag{Name: "docker-cpu-period", Usage: "Set docker CPU period NOTIMPLEMENTED", Hidden: true},
		cli.IntFlag{Name: "docker-cpu-quota", Usage: "Set docker CPU quota NOTIMPLEMENTED", Hidden: true},
		cli.IntFlag{Name: "docker-memory", Usage: "Set docker user memory limit in MB", Hidden: true},
		cli.IntFlag{Name: "docker-memory-swap", Usage: "Set docker user memory swap limit in MB", Hidden: true},
		cli.IntFlag{Name: "docker-memory-reservation", Usage: "Set docker user memory soft limit in MB NOTIMPLEMENTED", Hidden: true},
		cli.IntFlag{Name: "docker-kernel-memory", Usage: "Set docker kernel memory limit in MB NOTIMPLEMENTED", Hidden: true},
		cli.BoolFlag{Name: "docker-cleanup-image", Usage: "Remove image from the Docker when finished pushing them", Hidden: true},
		cli.StringFlag{Name: "docker-network", Value: "", Usage: "Docker network name.", Hidden: true},
	}

	// These flags control where we store local files
	LocalPathFlags = []cli.Flag{
		cli.StringFlag{Name: "working-dir", Value: "./.wercker", Usage: "Path where we store working files.", EnvVar: "WERCKER_WORKING_DIR"},
	}

	// These flags control paths on the guest and probably shouldn't change
	InternalPathFlags = []cli.Flag{
		cli.StringFlag{Name: "mnt-root", Value: "/mnt", Usage: "Directory on the guest where volumes are mounted.", Hidden: true},
		cli.StringFlag{Name: "guest-root", Value: "/pipeline", Usage: "Directory on the guest where work is done.", Hidden: true},
		cli.StringFlag{Name: "report-root", Value: "/report", Usage: "Directory on the guest where reports will be written.", Hidden: true},
	}

	// These flags are usually pulled from the env
	WerckerFlags = []cli.Flag{
		cli.StringFlag{Name: "build-id", Value: "", EnvVar: "WERCKER_BUILD_ID", Hidden: true, Usage: "The build id (deprecated)."},
		cli.StringFlag{Name: "deploy-id", Value: "", EnvVar: "WERCKER_DEPLOY_ID", Hidden: true, Usage: "The deploy id (deprecated)."},
		cli.StringFlag{Name: "run-id", Value: "", EnvVar: "WERCKER_DEPLOY_ID", Hidden: true, Usage: "The run id."},
		cli.StringFlag{Name: "deploy-target", Value: "", EnvVar: "WERCKER_DEPLOYTARGET_NAME", Usage: "The deploy target name."},
		cli.StringFlag{Name: "application-id", Value: "", EnvVar: "WERCKER_APPLICATION_ID", Hidden: true, Usage: "The application id."},
		cli.StringFlag{Name: "application-name", Value: "", EnvVar: "WERCKER_APPLICATION_NAME", Hidden: true, Usage: "The application name."},
		cli.StringFlag{Name: "application-owner-name", Value: "", EnvVar: "WERCKER_APPLICATION_OWNER_NAME", Hidden: true, Usage: "The application owner name."},
		cli.StringFlag{Name: "application-started-by-name", Value: "", EnvVar: "WERCKER_APPLICATION_STARTED_BY_NAME", Hidden: true, Usage: "The name of the user who started the application."},
		cli.StringFlag{Name: "wercker-container-registry", Value: "https://wcr.io/v2/", Usage: "Wercker Registry Domain.", EnvVar: "WERCKER_CONTAINER_REGISTRY", Hidden: true},
		cli.StringFlag{Name: "pipeline", Value: "", EnvVar: "WERCKER_PIPELINE", Usage: "Alternate pipeline name to execute."},
	}

	GitFlags = []cli.Flag{
		cli.StringFlag{Name: "git-domain", Value: "", Usage: "Git domain.", EnvVar: "WERCKER_GIT_DOMAIN", Hidden: true},
		cli.StringFlag{Name: "git-owner", Value: "", Usage: "Git owner.", EnvVar: "WERCKER_GIT_OWNER", Hidden: true},
		cli.StringFlag{Name: "git-repository", Value: "", Usage: "Git repository.", EnvVar: "WERCKER_GIT_REPOSITORY", Hidden: true},
		cli.StringFlag{Name: "git-branch", Value: "", Usage: "Git branch.", EnvVar: "WERCKER_GIT_BRANCH", Hidden: true},
		cli.StringFlag{Name: "git-commit", Value: "", Usage: "Git commit.", EnvVar: "WERCKER_GIT_COMMIT", Hidden: true},
	}

	// These flags affect our registry interactions
	RegistryFlags = []cli.Flag{
		cli.StringFlag{Name: "commit", Value: "", Usage: "Commit the build result locally."},
		cli.StringFlag{Name: "tag", Value: "", Usage: "Tag for this build.", EnvVar: "WERCKER_GIT_BRANCH"},
		cli.StringFlag{Name: "message", Value: "", Usage: "Message for this build."},
	}

	// These flags affect our artifact interactions
	ArtifactFlags = []cli.Flag{
		cli.BoolFlag{Name: "artifacts", Usage: "Store artifacts."},
		cli.BoolFlag{Name: "no-remove", Usage: "Don't remove the containers."},
		cli.BoolFlag{Name: "store-s3",
			Usage: `Store artifacts and containers on s3.
			This requires access to aws credentials, pulled from any of the usual places
			(~/.aws/config, AWS_SECRET_ACCESS_KEY, etc), or from the --aws-secret-key and
			--aws-access-key flags. It will upload to a bucket defined by --s3-bucket in
			the region named by --aws-region`},
	}

	// These flags affect our local execution environment
	DevFlags = []cli.Flag{
		cli.StringFlag{Name: "environment", Value: "ENVIRONMENT", Usage: "Specify additional environment variables in a file.", EnvVar: "WERCKER_ENVIRONMENT_FILE"},
		cli.BoolFlag{Name: "verbose", Usage: "Print more information."},
		cli.BoolFlag{Name: "no-colors", Usage: "Wercker output will not use colors (does not apply to step output)."},
		cli.BoolFlag{Name: "debug", Usage: "Print additional debug information."},
		cli.BoolFlag{Name: "journal", Usage: "Send logs to systemd-journald. Suppresses stdout logging."},
	}

	// These flags are advanced dev settings
	InternalDevFlags = []cli.Flag{
		cli.BoolTFlag{Name: "direct-mount", Usage: "Mount our binds read-write to the pipeline path."},
		cli.BoolFlag{Name: "expose-ports", Usage: "Enable ports from wercker.yml being exposed to the host system."},
		// deprecated
		cli.StringSliceFlag{Name: "publish", Value: &cli.StringSlice{}, Usage: "[Deprecated] Use: --expose-ports. - Publish a port from the main container, same format as docker --publish.", Hidden: true},
		cli.BoolFlag{Name: "attach-on-error", Usage: "Attach shell to container if a step fails.", Hidden: true},
		cli.BoolFlag{Name: "enable-volumes", Usage: "Mount local files and directories as volumes to your wercker container, specified in your wercker.yml."},
		cli.BoolTFlag{Name: "enable-dev-steps", Hidden: true, Usage: `
		Enable internal dev steps.
		This enables:
		- internal/watch
		`},
	}

	// These flags are advanced build settings
	InternalBuildFlags = []cli.Flag{
		cli.BoolFlag{Name: "direct-mount", Usage: "Mount our binds read-write to the pipeline path."},
		cli.BoolFlag{Name: "expose-ports", Usage: "Enable ports from wercker.yml being exposed to the host system."},
		// deprecated
		cli.StringSliceFlag{Name: "publish", Value: &cli.StringSlice{}, Usage: "[Deprecated] Use: --expose-ports. - Publish a port from the main container, same format as docker --publish.", Hidden: true},
		cli.BoolFlag{Name: "attach-on-error", Usage: "Attach shell to container if a step fails.", Hidden: true},
		cli.BoolFlag{Name: "enable-volumes", Usage: "Mount local files and directories as volumes to your wercker container, specified in your wercker.yml."},
		cli.BoolFlag{Name: "enable-dev-steps", Hidden: true, Usage: `
		Enable internal dev steps.
		This enables:
		- internal/watch
		`},
	}

	// Flags for advanced deploy settings
	InternalDeployFlags = []cli.Flag{
		cli.BoolFlag{Name: "expose-ports", Usage: "Enable ports from wercker.yml being exposed to the host system."},
		// deprecated
		cli.StringSliceFlag{Name: "publish", Value: &cli.StringSlice{}, Usage: "[Deprecated] Use: --expose-ports. - Publish a port from the main container, same format as docker --publish.", Hidden: true},
		cli.BoolFlag{Name: "attach-on-error", Usage: "Attach shell to container if a step fails.", Hidden: true},
		cli.BoolFlag{Name: "enable-dev-steps", Hidden: true, Usage: `
		Enable internal dev steps.
		This enables:
		- internal/watch
		`},
	}

	// AWS bits
	AWSFlags = []cli.Flag{
		cli.StringFlag{Name: "aws-secret-key", Value: "", Usage: "Secret access key. Used for artifact storage."},
		cli.StringFlag{Name: "aws-access-key", Value: "", Usage: "Access key id. Used for artifact storage."},
		cli.StringFlag{Name: "s3-bucket", Value: "wercker-development", Usage: "Bucket for artifact storage."},
		cli.StringFlag{Name: "aws-region", Value: "us-east-1", Usage: "AWS region to use for artifact storage."},
	}

	// Wercker Reporter settings
	ReporterFlags = []cli.Flag{
		cli.BoolFlag{Name: "report", Usage: "Report logs back to wercker (requires build-id, wercker-host, wercker-token).", Hidden: true},
		cli.StringFlag{Name: "wercker-host", Usage: "Wercker host to use for wercker reporter.", Hidden: true},
		cli.StringFlag{Name: "wercker-token", Usage: "Wercker token to use for wercker reporter.", Hidden: true},
	}

	// These options might be overwritten by the wercker.yml
	ConfigFlags = []cli.Flag{
		cli.StringFlag{Name: "ignore-file", Value: ".werckerignore", Usage: "File with file patterns to ignore when copying files."},
		cli.StringFlag{Name: "source-dir", Value: "", Usage: "Source path relative to checkout root."},
		cli.Float64Flag{Name: "no-response-timeout", Value: 5, Usage: "Timeout if no script output is received in this many minutes."},
		cli.Float64Flag{Name: "command-timeout", Value: 25, Usage: "Timeout if command does not complete in this many minutes."},
		cli.StringFlag{Name: "wercker-yml", Value: "", Usage: "Specify a specific yaml file.", EnvVar: "WERCKER_YML_FILE"},
	}

	// Steps options
	StepPublishFlags = []cli.Flag{
		cli.StringFlag{Name: "owner", Value: "", Usage: "owner of the step, leave blank to use the token owner"},
	}

	PullFlagSet = [][]cli.Flag{
		[]cli.Flag{
			cli.StringFlag{Name: "branch", Value: "", Usage: "Filter on this branch."},
			cli.StringFlag{Name: "result", Value: "", Usage: "Filter on this result (passed or failed)."},
			cli.StringFlag{Name: "output", Value: "./repository.tar", Usage: "Path to repository."},
			cli.BoolFlag{Name: "load", Usage: "Load the container into docker after downloading."},
			cli.BoolFlag{Name: "f, force", Usage: "Override output if it already exists."},
		},
	}

	GlobalFlagSet = [][]cli.Flag{
		DevFlags,
		EndpointFlags,
		AuthFlags,
	}

	DockerFlagSet = [][]cli.Flag{
		DockerFlags,
	}

	PipelineFlagSet = [][]cli.Flag{
		LocalPathFlags,
		WerckerFlags,
		DockerFlags,
		InternalBuildFlags,
		GitFlags,
		RegistryFlags,
		ArtifactFlags,
		AWSFlags,
		ConfigFlags,
	}

	DeployPipelineFlagSet = [][]cli.Flag{
		LocalPathFlags,
		WerckerFlags,
		DockerFlags,
		InternalDeployFlags,
		GitFlags,
		RegistryFlags,
		ArtifactFlags,
		AWSFlags,
		ConfigFlags,
	}

	DevPipelineFlagSet = [][]cli.Flag{
		LocalPathFlags,
		WerckerFlags,
		DockerFlags,
		InternalDevFlags,
		GitFlags,
		RegistryFlags,
		ArtifactFlags,
		AWSFlags,
		ConfigFlags,
	}

	WerckerInternalFlagSet = [][]cli.Flag{
		InternalPathFlags,
		ReporterFlags,
	}

	WerckerDockerFlagSet = [][]cli.Flag{
		AuthFlags,
		WerckerFlags,
	}

	ExternalRunnerCommonFlags = []cli.Flag{
		cli.StringFlag{Name: "name", Usage: "specify a unique name for active runner(s))"},
		cli.StringFlag{Name: "group", Usage: "specify the group name for starting runner(s)"},
	}

	ExternalRunnerStartFlags = []cli.Flag{
		cli.StringFlag{Name: "name", Usage: "specify an instance name for runner(s)"},
		cli.StringFlag{Name: "group", Usage: "specify the group name for runner(s)"},
		cli.StringFlag{Name: "orgs", Usage: "comma separated list of organization names"},
		cli.StringFlag{Name: "apps", Usage: "comma separated list of application names, format: org-name/app-name for each. "},
		cli.StringFlag{Name: "workflows", Usage: "comma separated list of workflows, format: org-name/app-name/workflow-name for each."},
		cli.StringFlag{Name: "storepath", Usage: "local file system path for storing runner output"},
		cli.StringFlag{Name: "logpath", Usage: "local file system path for storing log files"},
		cli.IntFlag{Name: "runners", Value: 1, Usage: "number of runners to start, default is one"},
		cli.IntFlag{Name: "poll-frequency", Value: 15, Usage: "number of seconds between runner polling for a job"},
		cli.StringFlag{Name: "token", Usage: "bearer token for external runner", EnvVar: "WERCKER_RUNNER_TOKEN"},
		cli.BoolFlag{Name: "all", Usage: "specify that all jobs allowed to the user are eligible for selection by this runner"},
	}
)

func FlagsFor(flagSets ...[][]cli.Flag) []cli.Flag {
	all := []cli.Flag{}
	for _, flagSet := range flagSets {
		for _, x := range flagSet {
			all = append(all, x...)
		}
	}
	return all
}
